shared/src/main/scala/mlscript/syntax.scala

```diff
@@ -19,7 +22,9 @@ final case class TypeDef(
   mthDecls: List[MethodDef[Right[Term, Type]]],
   mthDefs: List[MethodDef[Left[Term, Type]]],
   positionals: Ls[Var],
-) extends Decl
+) extends Decl {
+  var adtInfo: Opt[AdtInfo] = N
```

Comment:
Does this need to be mutable? Note that you can parameterize parser definitions with arbitrary parameters, such as passing in an `Option[AdtInfo]` when parsing cases.